From d3783a13f0dd9f606a940b3d2f0f06d775955941 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 10:12:49 -0400
Subject: [fs] ext4: use EXT_MAX_BLOCKS in ext4_es_can_be_merged()

Message-id: <1414059176-6044-26-git-send-email-lczerner@redhat.com>
Patchwork-id: 98302
O-Subject: [RHEL 7.1 PATCH 25/32] ext4: use EXT_MAX_BLOCKS in ext4_es_can_be_merged()
Bugzilla: 1150171

BZ 1150171
https://bugzilla.redhat.com/show_bug.cgi?id=1150171

BREW 8149772
https://brewweb.devel.redhat.com/taskinfo?taskID=8149772

Upstream commit 0baaea64009d5edbfee36a949017f7c68926fafc

 In ext4_es_can_be_merged() when checking whether we can merge two
 extents we should use EXT_MAX_BLOCKS instead of defining it manually.
 Also if it is really the case we should notify userspace because clearly
 there is a bug in extent status tree implementation since this should
 never happen.

 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Reviewed-by: Zheng Liu <wenqing.lz@taobao.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/extents_status.c b/fs/ext4/extents_status.c
index b1d7beb..e78af3a 100644
--- a/fs/ext4/extents_status.c
+++ b/fs/ext4/extents_status.c
@@ -344,8 +344,14 @@ static int ext4_es_can_be_merged(struct extent_status *es1,
  if (ext4_es_status(es1) != ext4_es_status(es2))
   return 0;
 
- if (((__u64) es1->es_len) + es2->es_len > 0xFFFFFFFFULL)
+ if (((__u64) es1->es_len) + es2->es_len > EXT_MAX_BLOCKS) {
+  pr_warn("ES assertion failed when merging extents. "
+   "The sum of lengths of es1 (%d) and es2 (%d) "
+   "is bigger than allowed file size (%d)\n",
+   es1->es_len, es2->es_len, EXT_MAX_BLOCKS);
+  WARN_ON(1);
   return 0;
+ }
 
  if (((__u64) es1->es_lblk) + es1->es_len != es2->es_lblk)
   return 0;
-- 
1.7.1